### DISCUSSION TITLE
Using CppAD_py interface in cygwin

### DISCUSSION THREAD

**Discussion by Manushanker** [coin-or/CppAD#101] [Q&A]
I am trying to compile cppad_py library in windows using cygwin. I followed the instructions described in the documentation https://bradbell.github.io/cppad_py/doc/xsrst/cppad_py.html, but could not succeed in installing it. May I know if it is possible to compile cppad_py in windows and use it in visual studio?

**Comment by bradbell** [comment]
Looking at my virtual machines, I have been testing CppAD in visual studio and msys2.
I am going to try installing cppad_py in msys2. Have you tired that ?

**Comment by Manushanker** [comment]
When I ran "python3 setup.py install --prefix=$prefix" from the main installation directory "cppadpyMain", I get the following error message

![image](https://user-images.githubusercontent.com/80901847/111637632-00205b80-87d0-11eb-846e-d324fa685f64.png)

Can you please let me know how to resolve this error message?

**Reply by bradbell** [reply]
Did you check the settings in `bin/get_cppad.sh` and run it first ?

The error you have looks very strange (it is as if bash is not working). I will have to install cygwin to see what happens for me.

**Reply by Manushanker** [reply]
I checked the settings in bin/get_cppad.sh but it not change anything. When I ran  bin/get_cppad.sh  I get this error message

manushanker@DESKTOP-F38DJ9R /cygdrive/c/cppadpyMain
$ bin/get_cppad.sh
bin/get_cppad.sh: line 127: syntax error near unexpected token `$'{\r''

**Reply by Manushanker** [reply]
Sorry I could understand the section "https://bradbell.github.io/cppad_py/doc/xsrst/get_cppad_sh.html#get-cppad-sh-settings". I have used CppAD C++ library in the past and since it is a header based library I could easily use it visual studio

**Reply by bradbell** [reply]
Now worries. I am installing cygwin on a virtual machine and will see what happens. It may take me a day or so to get back to you.

**Reply by Manushanker** [reply]
Thanks for your time and efforts!!

**Reply by Manushanker** [reply]
The main reason for the "syntax error near unexpected token '$'{\r" when running the command python3 setup.py install --prefix=$prefix is that the build_type.sh file when downloaded using git clone has a windows specific format. Which is different from unix file format, the issue got resolved when I used the package dos2unix to convert the files into Unix format.

**Reply by Manushanker** [reply]
After following the instructions mentioned in https://bradbell.github.io/cppad_py/doc/xsrst/setup_py.html#setup-py-prefix I managed to compile cppad_py library in cygwin!! But when I ran the command "python3 example/python/check_all.py
" I get the following error message 

![image](https://user-images.githubusercontent.com/80901847/111719611-9ab18680-8832-11eb-873d-b77d50258c76.png)

Do you know how to add cppad_py library in visual studio project?

**Comment by bradbell** [comment]
I am getting the same error:
```
bradl@DESKTOP-HG8IMEN ~/repo/cppad_py.git
$ python3 example/python/check_all.py
Traceback (most recent call last):
  File "example/python/check_all.py", line 90, in <module>
    ok = run_test(name, name)
  File "example/python/check_all.py", line 19, in run_test
    exec( 'ok = '   + module_name + '.' + fun_name + '()', namespace )
  File "<string>", line 1, in <module>
  File "./a_double_cond_assign_xam.py", line 14, in a_double_cond_assign_xam
    import cppad_py
  File "/home/bradl/repo/cppad_py.git/cppad_py/__init__.py", line 13, in <module>
    from cppad_py.cppad_swig import abort_recording
  File "/home/bradl/repo/cppad_py.git/cppad_py/cppad_swig.py", line 13, in <module>
    from . import _cppad_swig
ImportError: No such file or directory
```
For some reason on cygwin it is not importing the file
```
cppad_py/_cppad_swig.cpython-38-x86_64-cygwin.dll 
```
There is something about python extensions on cygwin that is failing here and needs to be tracked down.

**Comment by bradbell** [comment]
My simple swig test case does not work on cygwin. Try it and see what happens for you. Place the following files in an empty directory called swig and then execute the bash script `python.sh`. (You will probably have to change the version of python in `python.sh`)

my_example.cpp:
```
/* File : my_example.c */
 
# include <ctime>
double My_variable = 3.0;

int fact(int n) {
	if (n <= 1) return 1;
	else return n*fact(n-1);
}

int my_mod(int x, int y) {
	return (x%y);
}
	
char *get_time()
{
	time_t ltime;
	time(&ltime);
	return ctime(&ltime);
}
```
my_example.i:
```
/* my_example.i */
 %module my_example
 %{
 /* Put header files here or function declarations like below */
 extern double My_variable;
 extern int fact(int n);
 extern int my_mod(int x, int y);
 extern char *get_time();
 %}
 
 extern double My_variable;
 extern int fact(int n);
 extern int my_mod(int x, int y);
 extern char *get_time();
```
python.sh:
```
#! /bin/bash -e
# -----------------------------------------------------------------------------
# bash function that echos and executes a command
echo_eval() {
	echo $*
	eval $*
}
# ---------------------------------------------------------------------------
# create new build directory
if [ -e build ]
then
	echo_eval rm -r build
fi
echo_eval mkdir build
echo_eval cp my_example.* build
echo_eval cd build
# create my_example_wrap.cpp
echo_eval swig -python -c++ my_example.i
echo_eval mv my_example_wrap.cxx my_example_wrap.cpp
# build module that is loadable by python
echo_eval g++ my_example.cpp my_example_wrap.cpp \
	-c \
	-fPIC \
	-I /usr/include/python3.9 
echo_eval g++ -shared my_example.o my_example_wrap.o \
	-o _my_example.so
# ---------------------------------------------------------------------------
# text module
cat << EOF > run.py
# load the modules
import sys
import my_example
# initialze exit status as OK
error_count = 0;
# check factor function
if my_example.fact(4) == 24 :
	print("my_example.fact: OK")
else :
	print("my_example.fact: Error")
	error_count = error_count + 1;
# check my_mod function
if my_example.my_mod(4,3) == 1 :
	print("my_example.my_mod: OK")
else :
	print("my_example.my_mod: Error")
	error_count = error_count + 1;
#
# return error_count 
sys.exit(error_count)
EOF
# ---------------------------------------------------------------------------
if python run.py
then
	echo 'All tests passed'
else
	echo 'At least one test failed'
fi
```

**Reply by Manushanker** [reply]
I ran it in my pc I get the following error message
![image](https://user-images.githubusercontent.com/80901847/111784224-851f7980-8891-11eb-850d-cdc2da31e6e9.png)

**Comment by bradbell** [comment]
Yes, that is the same as on my machine. It seems like this is a problem with using swig on cygwin. 
I have posted a question about it on the swig user's group; see
https://sourceforge.net/p/swig/mailman/message/37243840/

**Reply by Manushanker** [reply]
Thanks !!

**Comment by bradbell** [comment]
I have posted a swig issue; see
https://github.com/swig/swig/issues/1983
Perhaps this will get some response.

**Comment by bradbell** [comment]
Here is the solution for getting the tutorial to build:
https://cygwin.com/pipermail/cygwin/2021-March/248132.html

**Comment by bradbell** [comment]
The main issue here is that the CppAD dynamic library llib_cppad was not linking because its location was not in the `PATH` variable (on windows the `PATH` variable plays the role of the `LD_LIBRARY_PATH` variable on unix. In addition, dynamic libraries do not work with windows dlls because they handle static variables differently. This was fixed by building static versions of the cppad_lib on msys and cygwin; see 04-27 on
https://coin-or.github.io/CppAD/doc/whats_new_21.htm#04-27

This problem I have run into now is getting msys or cygwin to install scipy. Only numpy is really necessary, but scipy is used by some of the tests and utilities.

**Reply by Manushanker** [reply]
Thanks for your time and efforts to resolve this issue. Instead of using CppAD python library, I have created a wrapper code using Boost python library to call CppAD C++  from python. It works without any issue. Thanks one again for your help

**Comment by bradbell** [comment]
I installed the following using the cygwin installer
gcc-core, gcc-g++, lapack-devel,  
python3-devel, python-pip-wheel, python-setuptools-wheel,  
git, vim, libopenblas, gcc-fortran, cmake, make, swig 

The I installed to following using pip3
numpy, scipy

Then the build of cppad_py worked fine on cygwin

**Reply by Manushanker** [reply]
Thanks! I will try to do the same
